os/net/netdev/netdev_ioctl.c

```diff
@@ -1245,46 +1262,70 @@ void netdev_ifup(FAR struct netif *dev)
 {
 	/* Make sure that the device supports the d_ifup() method */
 
-	if (dev->d_ifup) {
-		/* Is the interface already up? */
+	if (!dev->d_ifup) {
+		return;
+	}
 
-		if ((dev->d_flags & IFF_UP) == 0) {
-			/* No, bring the interface up now */
+	/* Is the interface already up? */
+	if ((dev->d_flags & IFF_UP) != 0) {
+		/* No, bring the interface up now */
+		return;
+	}
 
-			if (dev->d_ifup(dev) == OK) {
-				/* Mark the interface as up */
+	if (dev->d_ifup(dev) != OK) {
+		ndbg("fail to ifup %s:%d\n", __FILE__, __LINE__);
+		return;
+	}
 
-				dev->d_flags |= IFF_UP;
-				/* netdev_ifdown is blocked function, but some API that is mapping d_ifup
-				 * operated as non-blocking, so disalbe netdev_ifdown temporarily until API is fixed
-				 */
-				sleep(3);
-				netif_set_up(dev);
-				netif_set_link_up(dev);
-			}
-		}
+	/* Mark the interface as up */
+	dev->d_flags |= IFF_UP;
+
+	/* netdev_ifdown is a blocked function, but some API that is mapping d_ifup
+	 * operated as non-blocking, so disalbe netdev_ifdown temporarily until API is fixed
+	 */
+	sleep(3);
```

Comment:
This three second sleeping just gives a rise to unnecessary overhead in case that dev->d_ifup is a blocked function.
So, why don't introduce a new config like 'CONFIG_NET_IF_UPDOWN_FUNCTION_TYPE'?
Only if this config is set to 'non-blocked', then sleep(3) would be introduced here.

Reply:
thanks for your opinion. let me explain history, currently ifup function of wi-fi of artik053 is non-block function. so sleep(3) is necessary. I agree that the new feature would lower overhead. But that feature is not related to this pull request. Therefore I'll make new commit for that.